nano/node/request_aggregator.cpp

```diff
@@ -0,0 +1,225 @@
+	nano::unique_lock<std::mutex> lock (mutex);
+	condition.wait (lock, [& started = started] { return started; });
+}
+// clang-format on
+
+void nano::request_aggregator::add (std::shared_ptr<nano::transport::channel> & channel_a, std::vector<std::pair<nano::block_hash, nano::root>> const & hashes_roots_a)
+{
+	assert (wallets.rep_counts ().voting > 0);
+	auto endpoint (nano::transport::map_endpoint_to_v6 (channel_a->get_endpoint ()));
+	nano::unique_lock<std::mutex> lock (mutex);
+	auto & requests_by_endpoint (requests.get<tag_endpoint> ());
+	auto existing (requests_by_endpoint.find (endpoint));
+	if (existing == requests_by_endpoint.end ())
+	{
+		existing = requests_by_endpoint.emplace (channel_a).first;
+	}
+	// clang-format off
+	requests_by_endpoint.modify (existing, [&hashes_roots_a, &channel_a, this](channel_pool & pool_a) {
+		// This extends the lifetime of the channel, which is acceptable up to max_delay
+		pool_a.channel = channel_a;
+		auto new_deadline (std::min (pool_a.start + this->max_delay, std::chrono::steady_clock::now () + this->small_delay));
+		pool_a.deadline = new_deadline;
+		for (auto const & hash_root : hashes_roots_a)
+		{
+			pool_a.hashes_roots.emplace_back (hash_root.first, hash_root.second);
+		}
+	});
+	// clang-format on
+	if (requests.size () == 1)
+	{
+		lock.unlock ();
+		condition.notify_all ();
+	}
+}
+
+void nano::request_aggregator::run ()
+{
+	nano::thread_role::set (nano::thread_role::name::request_aggregator);
+	nano::unique_lock<std::mutex> lock (mutex);
+	started = true;
+	lock.unlock ();
+	condition.notify_all ();
+	lock.lock ();
+	while (!stopped)
+	{
+		if (!requests.empty ())
+		{
+			auto & requests_by_deadline (requests.get<tag_deadline> ());
+			auto front (requests_by_deadline.begin ());
+			if (front->deadline < std::chrono::steady_clock::now ())
+			{
+				auto pool (*front);
+				auto transaction (store.tx_begin_read ());
+				// Aggregate the current pool of requests, sending cached votes
+				// Store a local copy of the remaining hashes and the channel
+				auto remaining = aggregate (transaction, pool);
+				auto channel = pool.channel;
+				// Safely erase this pool
+				requests_by_deadline.erase (front);
+				if (!remaining.empty ())
+				{
+					lock.unlock ();
+					// Generate votes for the remaining hashes
+					generate (transaction, std::move (remaining), channel);
+					lock.lock ();
+				}
+			}
+			else
+			{
+				auto deadline = front->deadline;
+				// clang-format off
+				condition.wait_until (lock, deadline, [this, &deadline]() { return this->stopped || deadline < std::chrono::steady_clock::now (); });
+				// clang-format on
+			}
+		}
+		else
+		{
+			// clang-format off
+			condition.wait_for (lock, small_delay, [this]() { return this->stopped || !this->requests.empty (); });
+			// clang-format on
+		}
+	}
+}
+
+void nano::request_aggregator::stop ()
+{
+	nano::unique_lock<std::mutex> lock (mutex);
+	stopped = true;
+
+	lock.unlock ();
+	condition.notify_all ();
```

Comment:
You can limit the lock scope and make use of the RAII pattern we use elsewhere i.e
```
{
  nano::lock_guard<std::mutex> guard (mutex);
  stopped = true;
}
condition.notify_all ();
```